src/patch_parse.c

```diff
@@ -0,0 +1,1025 @@
+			newlines--;
+			break;
+
+		default:
+			error = parse_err("invalid patch hunk at line %d", ctx->line_num);
+			goto done;
+		}
+
+		line = git_array_alloc(patch->base.lines);
+		GITERR_CHECK_ALLOC(line);
+
+		memset(line, 0x0, sizeof(git_diff_line));
+
+		line->content = ctx->line + prefix;
+		line->content_len = ctx->line_len - prefix;
+		line->content_offset = ctx->content_len - ctx->remain;
+		line->origin = origin;
+
+		hunk->line_count++;
+	}
+
+	if (oldlines || newlines) {
+		error = parse_err(
+			"invalid patch hunk, expected %d old lines and %d new lines",
+			hunk->hunk.old_lines, hunk->hunk.new_lines);
+		goto done;
+	}
+
+	/* Handle "\ No newline at end of file".  Only expect the leading
+	* backslash, though, because the rest of the string could be
+	* localized.  Because `diff` optimizes for the case where you
+	* want to apply the patch by hand.
+	*/
+	if (ctx->line_len >= 2 && memcmp(ctx->line, "\\ ", 2) == 0 &&
+		git_array_size(patch->base.lines) > 0) {
+
+		line = git_array_get(patch->base.lines, git_array_size(patch->base.lines) - 1);
+
+		if (line->content_len < 1) {
+			error = parse_err("cannot trim trailing newline of empty line");
+			goto done;
+		}
+
+		line->content_len--;
+
+		parse_advance_line(ctx);
+	}
+
+done:
+	return error;
+}
+
+static int parse_header_traditional(
+	git_patch_parsed *patch, patch_parse_ctx *ctx)
+{
+	GIT_UNUSED(patch);
+	GIT_UNUSED(ctx);
+
+	return 1;
+}
+
+static int parsed_patch_header(
+	git_patch_parsed *patch,
+	patch_parse_ctx *ctx)
+{
+	int error = 0;
+
+	for (ctx->line = ctx->content; ctx->remain > 0; parse_advance_line(ctx)) {
+		/* This line is too short to be a patch header. */
+		if (ctx->line_len < 6)
+			continue;
+
+		/* This might be a hunk header without a patch header, provide a
+		* sensible error message. */
+		if (memcmp(ctx->line, "@@ -", 4) == 0) {
+			size_t line_num = ctx->line_num;
+			git_patch_hunk hunk;
+
+			/* If this cannot be parsed as a hunk header, it's just leading
+			* noise, continue.
+			*/
+			if (parse_hunk_header(&hunk, ctx) < 0) {
+				giterr_clear();
+				continue;
+			}
+
+			error = parse_err("invalid hunk header outside patch at line %d",
+				line_num);
+			goto done;
+		}
+
+		/* This buffer is too short to contain a patch. */
+		if (ctx->remain < ctx->line_len + 6)
+			break;
+
+		/* A proper git patch */
+		if (ctx->line_len >= 11 && memcmp(ctx->line, "diff --git ", 11) == 0) {
+			error = parse_header_git(patch, ctx);
+			goto done;
+		}
+
+		if ((error = parse_header_traditional(patch, ctx)) <= 0)
+			goto done;
+
+		error = 0;
+		continue;
+	}
+
+	error = parse_err("no header in patch file");
+
+done:
+	return error;
+}
+
+static int parsed_patch_binary_side(
+	git_diff_binary_file *binary,
+	patch_parse_ctx *ctx)
+{
+	git_diff_binary_t type = GIT_DIFF_BINARY_NONE;
+	git_buf base85 = GIT_BUF_INIT, decoded = GIT_BUF_INIT;
+	git_off_t len;
+	int error = 0;
+
+	if (ctx->line_len >= 8 && memcmp(ctx->line, "literal ", 8) == 0) {
```

Comment:
In this area of the patches, I'm wondering that, since we've already checked that one side is long enough, why use `memcmp()` here instead of `strcmp()` which would allow us to remove the manual length specification. I admit I don't quite know how well compiler optimise the case for a static string, though, which might be a valid reason to do this.